src/Bicep.Core/Analyzers/Linter/Rules/UseRecentModuleVersionsRule.cs

```diff
@@ -0,0 +1,220 @@
+using Bicep.Core.CodeAction;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Navigation;
+using Bicep.Core.Parsing;
+using Bicep.Core.Registry.Oci;
+using Bicep.Core.Registry.PublicRegistry;
+using Bicep.Core.Resources;
+using Bicep.Core.Semantics;
+using Bicep.Core.Syntax;
+using Bicep.Core.Text;
+using Bicep.Core.TypeSystem;
+using Bicep.Core.Workspaces;
+using Json.Patch;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.WindowsAzure.ResourceStack.Common.Extensions;
+using Semver;
+using Semver.Comparers;
+
+namespace Bicep.Core.Analyzers.Linter.Rules
+{
+    public sealed class UseRecentModuleVersionsRule : LinterRuleBase
+    {
+        public new const string Code = "use-recent-module-versions";
+
+        public record Failure(
+            TextSpan Span,
+            string Message,
+            string[] AcceptableVersions,
+            CodeFix[] Fixes
+        );
+
+        public UseRecentModuleVersionsRule() : base(
+            code: Code,
+            description: CoreResources.UseRecentModuleVersionsRule_Description,
+            LinterRuleCategory.BestPractice,
+            docUri: new Uri($"https://aka.ms/bicep/linter/{Code}"),
+            overrideCategoryDefaultDiagnosticLevel: DiagnosticLevel.Off // many users prefer this to be off by default due to the noise
+            )
+        {
+        }
+
+        public override string FormatMessage(params object[] values)
+        {
+            var message = (string)values[0];
+            var acceptableVersions = (string[])values[1];
+
+            var acceptableVersionsString = string.Join(", ", acceptableVersions);
+            return message
+                + (acceptableVersionsString.Length > 0
+                    // Currently we only support using the most recent version
+                    ? " " + string.Format(CoreResources.UseRecentModulesVersionRule_MostRecentVersion, acceptableVersionsString)
+                    : "");
+        }
+
+        override public IEnumerable<IDiagnostic> AnalyzeInternal(SemanticModel model, IServiceProvider serviceProvider, DiagnosticLevel diagnosticLevel)
+        {
+            return GetFailures(model, serviceProvider, diagnosticLevel)
+                        .Select(f => CreateFixableDiagnosticForSpan(diagnosticLevel, f.Span, f.Fixes, f.Message, f.AcceptableVersions));
+        }
+
+        private IEnumerable<Failure> GetFailures(SemanticModel model, IServiceProvider serviceProvider, DiagnosticLevel diagnosticLevel)
+        {
+            var publicRegistryModuleMetadataProvider = serviceProvider.GetRequiredService<IPublicRegistryModuleMetadataProvider>();
+            var hasShownDownloadWarning = false;
+
+            foreach (var (syntax, artifactResolutionInfo) in model.Compilation.SourceFileGrouping.ArtifactLookup
+                .Where(entry => entry.Value.Origin == model.SourceFile))
+            {
+                if (syntax is ModuleDeclarationSyntax moduleSyntax)
```

Comment:
It could be beneficial to combine the if condition with the Where clause to avoid looping twice.

Reply:
Don't see why it would loop twice.  Tried combining, but it's more complicated because you lose the type restraint on syntax is ModuleDeclarationSyntax.